src/rust/engine/process_execution/src/local.rs

```diff
@@ -452,9 +460,56 @@ pub trait CapturedWorkdir {
       .compat()
       .await?
     };
-    if let Some(workdir) = maybe_workdir {
-      // Dropping the temporary directory will likely involve a lot of IO: do it in the background.
-      let _background_cleanup = executor.spawn_blocking(|| std::mem::drop(workdir));
+
+    match maybe_workdir {
+      Some(workdir) => {
+        // Dropping the temporary directory will likely involve a lot of IO: do it in the
+        // background.
+        let _background_cleanup = executor.spawn_blocking(|| std::mem::drop(workdir));
+      }
+      None => {
+        // If we don't cleanup the workdir, we materialize a file named `__run.sh` into the output
```

Comment:
Will this script always be used to execute the command? Or is it just being written for later use in debugging if the workdir is preserved? If the latter, then consider writing the file with variables and arguments on separate lines to help future readers.

Reply:
This script will not be used to execute the command -- in fact, it is not present in the chroot until the command has completed. It would make sense to have env vars and arguments on separate lines -- I'll make that change.

Reply:
Hm. Without using it to execute the command (which seems reasonable, I think, as it avoids including a generated file in our error output in the common case), it's pretty likely to go out of sync with how we're actually executing things.

Any thoughts on how we can avoid this going out of sync? I don't think it's a blocker, as I agree that this is useful.

Reply:
Yes, I also considered this but didn't serialize my thoughts. I thought of using `__run.sh` to actually execute the process, but I *really* didn't want to suddenly use a different code path (via `__run.sh`) by toggling on debugging with `--no-process-execution-cleanup-local-dirs`.

To me it seemed like the least surprising thing if things did get out of sync, if the `__run.sh` script was wrong. At least we would have the same actual process execution with `--no-process-execution-cleanup-local-dirs` still doing the same code path as not. By keeping `__run.sh` as "purely" debug info, it seems to make the worst possible thing that could happen that our debugging tool gets out of sync, as opposed to having silent errors by doing process execution itself a little differently.

Does that make sense? I'm not fully satisfied with this solution, and I think separately from the remexec API itself, we as pants could adopt a policy of always performing remote executions with such a `__run.sh` file, if we wanted to. I definitely think I would prefer to let `__run.sh` leak to remote executions than having processes executed differently locally and remotely.